sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala

```diff
@@ -0,0 +1,222 @@
+import special.sigma.Context
+
+/** Configuration parameters of the evaluation run. */
+case class EvalSettings(
+  /** Used by [[ErgoTreeEvaluator.profiler]] to measure operations timings. */
+  isMeasureOperationTime: Boolean,
+  /** Used by [[ErgoTreeEvaluator]] to conditionally perform debug mode operations. */
+  isDebug: Boolean = false,
+  /** Used by [[ErgoTreeEvaluator]] to conditionally emit log messages. */
+  isLogEnabled: Boolean = false)
+
+/** Implements a Simple Fast direct-style interpreter of ErgoTrees.
+  *
+  * ### Motivation
+  * [[ErgoTree]] is a simple declarative intermediate representation for Ergo contracts. It is
+  * designed to be compact in serialized form and directly executable, i.e. no additional
+  * transformation is necessary before it can be efficiently executed.
+  *
+  * This class implements a big-step recursive interpreter that works directly with
+  * ErgoTree HOAS. Because of this the evaluator is very simple and follows denotational
+  * semantics of ErgoTree (see https://ergoplatform.org/docs/ErgoTree.pdf). Or, the other
+  * way around, this implementation of ErgoTreeEvaluator is purely functional with
+  * immutable data structures and can be used as definition of ErgoTree's semantics.
+  *
+  * ### Implementation
+  * ErgoTreeEvaluator takes ErgoTree directly as it is deserialized as part of a
+  * transaction. No additional transformation is performed.
+  * ErgoTree is interpreted directly and all the intermediate data is stored in the
+  * runtime types.
+  * The runtime types are such types as [[special.collection.Coll]],
+  * [[special.sigma.SigmaProp]], [[special.sigma.AvlTree]], [[BigInt]], etc.
+  * It also use immutable Map to keep current [[DataEnv]] of computed [[ValDef]]s, as
+  * result only addition is used from the map, and deletion is essentially a garbage
+  * collection.
+  *
+  * ### Performance
+  * Since this interpreter directly works with SigmaDsl types (Coll, BigInt, SigmaProp
+  * etc), it turns out to be very fast. Since it also does JIT style costing instead of
+  * AOT style, it is 5-6x faster than existing implementation.
+  *
+  * @param context   Represents blockchain data context for ErgoTree evaluation
+  * @param constants Segregated constants from ErgoTree, to lookup them from
+  *                  [[ConstantPlaceholder]] evaluation.
+  * @param coster    Accumulates computation costs.
+  * @param profiler  Performs operations profiling and time measurements (if enabled in settings).
+  * @param settings  Settings to be used during evaluation.
+  */
+class ErgoTreeEvaluator(
+  val context: Context,
+  val constants: Seq[Constant[SType]],
+  val coster: CostAccumulator,
+  val profiler: Profiler,
+  val settings: EvalSettings)
+{
+  /** Evaluates the given expression in the given data environment. */
+  def eval(env: DataEnv, exp: SValue): Any = {
+    exp.evalTo[Any](this, env)
+  }
+
+  def addCostOf(opName: String, opType: SFunc) = {
```

Comment:
please do ScalaDoc for all the public methods

Reply:
done